modules/data-streams/src/main/java/org/elasticsearch/datastreams/action/TransportReindexDataStreamIndexAction.java

```diff
@@ -0,0 +1,214 @@
+        logger.info("Creating destination index [{}] for source index [{}]", destIndexName, sourceIndex.getIndex().getName());
+
+        // Create destination with subset of source index settings that can be added before reindex
+        var settings = getPreSettings(sourceIndex);
+
+        var sourceMapping = sourceIndex.mapping();
+        Map<String, Object> mapping = sourceMapping != null ? sourceMapping.rawSourceAsMap() : Map.of();
+        var createIndexRequest = new CreateIndexRequest(destIndexName).settings(settings).mapping(mapping);
+
+        var errorMessage = String.format(Locale.ROOT, "Could not create index [%s]", destIndexName);
+        client.admin().indices().create(createIndexRequest, failIfNotAcknowledged(listener, errorMessage));
+    }
+
+    private void reindex(String sourceIndexName, String destIndexName, ActionListener<BulkByScrollResponse> listener) {
+        logger.info("Reindex to destination index [{}] from source index [{}]", destIndexName, sourceIndexName);
+        var reindexRequest = new ReindexRequest();
+        reindexRequest.setSourceIndices(sourceIndexName);
+        reindexRequest.getSearchRequest().allowPartialSearchResults(false);
+        reindexRequest.getSearchRequest().source().fetchSource(true);
+        reindexRequest.setDestIndex(destIndexName);
+        client.execute(ReindexAction.INSTANCE, reindexRequest, listener);
+    }
+
+    private void updateSettings(IndexMetadata sourceIndex, String destIndexName, ActionListener<AcknowledgedResponse> listener) {
+        logger.info("Adding settings from source index that could not be added before reindex");
+
+        Settings postSettings = getPostSettings(sourceIndex);
+        if (postSettings.isEmpty()) {
+            listener.onResponse(null);
+            return;
+        }
+
+        var updateSettingsRequest = new UpdateSettingsRequest(postSettings, destIndexName);
+        var errorMessage = String.format(Locale.ROOT, "Could not update settings on index [%s]", destIndexName);
+        client.admin().indices().updateSettings(updateSettingsRequest, failIfNotAcknowledged(listener, errorMessage));
+    }
+
+    // Filter source index settings to subset of settings that can be included during reindex
+    // TODO does tsdb start/end time get set in create request if copied into settings from source index
+    private Settings getPreSettings(IndexMetadata sourceIndex) {
+        return sourceIndex.getSettings().filter(settingName -> DISALLOWED_SETTINGS.contains(settingName) == false);
+    }
+
+    private Settings getPostSettings(IndexMetadata sourceIndex) {
+        // TODO
+        return Settings.EMPTY;
+        // return sourceIndex.getSettings().filter(DISALLOWED_SETTINGS::contains);
+    }
+
+    public static String generateDestIndexName(String sourceIndex) {
+        return "upgrade-" + sourceIndex;
```

Comment:
Is there logic anywhere that assumes a data stream index name starts with `.ds-`? Do we need to put this `upgrade-` in the middle of the name instead?

Reply:
Yeah, very likely needs something more complicated than this

Reply:
Surprisingly, this actually appears to be fine. Searchable snaphot indices that are swapped in actually are in this format, like `restored-.ds ....`